RowData.GetUint64() crashes on uint64 value

dml_events.go contains a helper method for converting MySQL values to `uint64`:

```
func (r RowData) GetUint64(colIdx int) (res uint64, err error) {
```

which documents that MySQL **never** returns `uint64` values. In experiments I see that for certain values, however, this is not the case, and invoking the method with `RowData` that already contains the expected value type crashes

I will try to reproduce the issue using a unit-test, unfortunately I lost the stack-trace where this was happening, but I know that the crash happened on a row of table

```
CREATE TABLE `mytable` (
  `mytable_id` bigint(19) unsigned NOT NULL AUTO_INCREMENT,
...
  PRIMARY KEY (`entry_info_id`),
...
) ENGINE=InnoDB AUTO_INCREMENT=13871426091 DEFAULT CHARSET=utf8
```

and an `mytable_id` value of `13871229209` .

unfortunately just ran into another stacktrace, even with my patch:

```
panic: reflect: call of reflect.Value.Int on uint32 Value

goroutine 11 [running]:
reflect.Value.Int(...)
	/usr/lib/go-1.14/src/reflect/value.go:976
github.com/Shopify/ghostferry.RowData.GetUint64(0xc0027f6990, 0x3, 0x3, 0x0, 0x0, 0x0, 0x1)
	/work/go/home/src/github.com/Shopify/ghostferry/dxl_events.go:32 +0x31d
github.com/Shopify/ghostferry.paginationKeyFromEventData(0xc0000d9940, 0xc0027f6990, 0x3, 0x3, 0x0, 0x0, 0x0)
	/work/go/home/src/github.com/Shopify/ghostferry/dxl_events.go:600 +0xdf
github.com/Shopify/ghostferry.(*BinlogDeleteEvent).PaginationKey(0xc0029e9860, 0xc0027f6c30, 0xc0004952b0, 0x10)
	/work/go/home/src/github.com/Shopify/ghostferry/dxl_events.go:245 +0x4c
github.com/Shopify/ghostferry.(*InlineVerifier).binlogEventListener(0xc00043b950, 0xc002e98370, 0xc000019828, 0xc002e98370)
	/work/go/home/src/github.com/Shopify/ghostferry/inline_verifier.go:602 +0x338
github.com/Shopify/ghostferry.(*BinlogStreamer).emitEvent(0xc000246500, 0xc0027f6c30, 0x1d, 0xc000019a48)
	/work/go/home/src/github.com/Shopify/ghostferry/binlog_streamer.go:347 +0x319
github.com/Shopify/ghostferry.(*BinlogStreamer).Run(0xc000246500)
	/work/go/home/src/github.com/Shopify/ghostferry/binlog_streamer.go:209 +0x569
github.com/Shopify/ghostferry.(*Ferry).Run.func7(0xc00003c200, 0xc00022e000)
	/work/go/home/src/github.com/Shopify/ghostferry/ferry.go:619 +0x5b
created by github.com/Shopify/ghostferry.(*Ferry).Run
	/work/go/home/src/github.com/Shopify/ghostferry/ferry.go:616 +0x2e0
```

it seems that we're not only getting `uint64` but any of the unsigned ints that we need to handle

--- a/dml_events.go
+++ b/dml_events.go
@@ -3,7 +3,6 @@ package ghostferry
 import (
 	"errors"
 	"fmt"
-	"reflect"
 	"regexp"
 	"strconv"
 	"strings"
@@ -19,24 +18,54 @@ var annotationRegex = regexp.MustCompile(`^/\*(.*?)\*/`)
 
 type RowData []interface{}
 
-// The mysql driver never actually gives you a uint64 from Scan, instead you
-// get an int64 for values that fit in int64 or a byte slice decimal string
-// with the uint64 value in it.
-func (r RowData) GetUint64(colIdx int) (res uint64, err error) {
+// For historical reasons, this function ended up being used at two different
+// places: the DataIterator and the DMLEvent (which indirectly is used by the
+// BinlogStreamer, InlineVerifier, etc).
+//
+// The original code here was introduced in
+// 152caec0ff5195d4698672df6dc0f72fb77b02fc, where it is used solely in context
+// of the DataIterator. In this context, the value being parsed here is given to
+// us by the go-sql-driver/mysql driver. This value could either by int64 or it
+// could be a byte slice decimal string with the uint64 value in it, which is
+// why we have this awkward byte slice to integer trick. This also means the
+// original code is not designed to handle uint64, as go-sql-driver/mysql never
+// returns uint64. This could possibly be an upstream problem that have since
+// been fixed, but this was not investigated.
+// (A possibly related PR: https://github.com/go-sql-driver/mysql/pull/690)
+//
+// At some point, this code was refactored into this function, such that the
+// BinlogStreamer also uses the same code to decode integers. The binlog data is
+// given to us by siddontang/go-mysql. The siddontang/go-mysql library should
+// not be giving us awkward byte slices. Instead, it should properly gives us
+// uint64. This code thus panics when it encounters such case. See
+// https://github.com/Shopify/ghostferry/issues/165.
+//
+// In summary:
+// - This code receives values from both go-sql-driver/mysql and
+//   siddontang/go-mysql.
+// - go-sql-driver/mysql gives us int64 for signed integer, and uint64 in a byte
+//   slice for unsigned integer.
+// - siddontang/go-mysql gives us int64 for signed integer, and uint64 for
+//   unsigned integer.
+// - We currently make this function deal with both cases. In the future we can
+//   investigate alternative solutions.
+func (r RowData) GetUint64(colIdx int) (uint64, error) {
+	u64, ok := Uint64Value(r[colIdx])
+	if ok {
+		return u64, nil
+	}
+
+	i64, ok := Int64Value(r[colIdx])
+	if ok {
+		return uint64(i64), nil
+	}
+
 	if valueByteSlice, ok := r[colIdx].([]byte); ok {
 		valueString := string(valueByteSlice)
-		res, err = strconv.ParseUint(valueString, 10, 64)
-		if err != nil {
-			return 0, err
-		}
-	} else {
-		signedInt := reflect.ValueOf(r[colIdx]).Int()
-		if signedInt < 0 {
-			return 0, fmt.Errorf("expected position %d in row to contain an unsigned number", colIdx)
-		}
-		res = uint64(signedInt)
+		return strconv.ParseUint(valueString, 10, 64)
 	}
-	return
+
+	return 0, fmt.Errorf("expected position %d to contain integer, but got type %T instead (with value of %v)", colIdx, r[colIdx], r[colIdx])
 }
 
 type DMLEvent interface {

--- a/test/go/trivial_integration_test.go
+++ b/test/go/trivial_integration_test.go
@@ -3,9 +3,10 @@ package test
 import (
 	sqlorig "database/sql"
 	"fmt"
-	sql "github.com/Shopify/ghostferry/sqlwrapper"
 	"math/rand"
 	"testing"
+
+	sql "github.com/Shopify/ghostferry/sqlwrapper"
 
 	"github.com/Shopify/ghostferry/testhelpers"
 )
@@ -67,20 +68,6 @@ func TestCopyDataWithDeleteLoad(t *testing.T) {
 			Tables:              []string{"gftest.table1"},
 		},
 		Ferry: testhelpers.NewTestFerry(),
-	}
-
-	testcase.Run()
-}
-
-func TestCopyDataWithLargePaginationKeyValues(t *testing.T) {
-	ferry := testhelpers.NewTestFerry()
-
-	ferry.Config.DataIterationBatchSize = 10
-
-	testcase := &testhelpers.IntegrationTestCase{
-		T:           t,
-		SetupAction: setupSingleTableDatabaseWithHighBitUint64PaginationKeys,
-		Ferry:       ferry,
 	}
 
 	testcase.Run()
@@ -171,11 +158,6 @@ func TestCopyDataWithNullInColumn(t *testing.T) {
 // Helper methods below
 // ====================
 
-func useUnsignedPaginationKey(db *sql.DB) {
-	_, err := db.Exec("ALTER TABLE gftest.table1 MODIFY id bigint(20) unsigned not null")
-	testhelpers.PanicIfError(err)
-}
-
 func setupSingleTableDatabase(f *testhelpers.TestFerry, sourceDB, targetDB *sql.DB) {
 	maxId := 1111
 	testhelpers.SeedInitialData(sourceDB, "gftest", "table1", maxId)
@@ -187,24 +169,6 @@ func setupSingleTableDatabase(f *testhelpers.TestFerry, sourceDB, targetDB *sql.
 	}
 
 	testhelpers.SeedInitialData(targetDB, "gftest", "table1", 0)
-}
-
-func setupSingleTableDatabaseWithHighBitUint64PaginationKeys(f *testhelpers.TestFerry, sourceDB, targetDB *sql.DB) {
-	setupSingleTableDatabase(f, sourceDB, targetDB)
-
-	_, err := sourceDB.Exec("TRUNCATE gftest.table1")
-	testhelpers.PanicIfError(err)
-
-	useUnsignedPaginationKey(sourceDB)
-	useUnsignedPaginationKey(targetDB)
-
-	stmt, err := sourceDB.Prepare("INSERT INTO gftest.table1 (id, data) VALUES (?, ?)")
-	testhelpers.PanicIfError(err)
-
-	for i := uint64(0); i < 100; i++ {
-		_, err = stmt.Exec(^uint64(0)-i, testhelpers.RandData())
-		testhelpers.PanicIfError(err)
-	}
 }
 
 func setupSingleTableDatabaseWithExtraNullColumn(f *testhelpers.TestFerry, sourceDB, targetDB *sql.DB) {

--- a/test/integration/types_test.rb
+++ b/test/integration/types_test.rb
@@ -276,20 +276,24 @@ class TypesTest < GhostferryTestCase
 
     ghostferry = new_ghostferry(MINIMAL_GHOSTFERRY, config: { verifier_type: "Inline" })
 
-    source_db.query("INSERT INTO #{DEFAULT_FULL_TABLE_NAME} (id) VALUES (12341234)")
+    max_uint64 = 18446744073709551615 # ^uint64(0)
+    expected = []
+    100.times do |i|
+      id = max_uint64 - i
+      expected << { "id" => id }
+      source_db.query("INSERT INTO #{DEFAULT_FULL_TABLE_NAME} (id) VALUES (#{id})")
+    end
 
     ghostferry.on_status(Ghostferry::Status::BINLOG_STREAMING_STARTED) do
       source_db.query("INSERT INTO #{DEFAULT_FULL_TABLE_NAME} (id) VALUES (72858243)")
     end
 
+    expected << { "id" => 72858243 }
+
     ghostferry.run
     assert_test_table_is_identical
 
-    expected = [
-      {"id" => 12341234},
-      {"id" => 72858248},
-    ]
-    res = target_db.query("SELECT * FROM #{DEFAULT_FULL_TABLE_NAME} ORDER BY id ASC")
+    res = target_db.query("SELECT * FROM #{DEFAULT_FULL_TABLE_NAME} ORDER BY id DESC")
 
     res.zip(expected).each do |row, expected_row|
       assert_equal expected_row, row

--- a/test/integration/types_test.rb
+++ b/test/integration/types_test.rb
@@ -268,6 +268,34 @@ class TypesTest < GhostferryTestCase
     end
   end
 
+  def test_unsigned_bigint_pk_with_inline_verifier
+    [source_db, target_db].each do |db|
+      db.query("CREATE DATABASE IF NOT EXISTS #{DEFAULT_DB}")
+      db.query("CREATE TABLE IF NOT EXISTS #{DEFAULT_FULL_TABLE_NAME} (id bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT, primary key(id))")
+    end
+
+    ghostferry = new_ghostferry(MINIMAL_GHOSTFERRY, config: { verifier_type: "Inline" })
+
+    source_db.query("INSERT INTO #{DEFAULT_FULL_TABLE_NAME} (id) VALUES (12341234)")
+
+    ghostferry.on_status(Ghostferry::Status::BINLOG_STREAMING_STARTED) do
+      source_db.query("INSERT INTO #{DEFAULT_FULL_TABLE_NAME} (id) VALUES (72858243)")
+    end
+
+    ghostferry.run
+    assert_test_table_is_identical
+
+    expected = [
+      {"id" => 12341234},
+      {"id" => 72858248},
+    ]
+    res = target_db.query("SELECT * FROM #{DEFAULT_FULL_TABLE_NAME} ORDER BY id ASC")
+
+    res.zip(expected).each do |row, expected_row|
+      assert_equal expected_row, row
+    end
+  end
+
   private
 
   def insert_json_on_source

